### PR TITLE
[FIXED] Error when importing an account results in an error, retry later

### PR DESCRIPTION
**Pull request by matthiashanel**
This failure can be caused if the imported account can't be fetched or
is not up to date yet.

Signed-off-by: Matthias Hanel <mh@synadia.com>

This change introduces a new account flag.
If this flag is set, perform the update and do NOT return ErrAccountResolverSameClaims.
Added two unit tests for both situations.
Shorten one unit test that was able to use the default operator.

### CODE REVIEW

**Comment by matthiashanel** [comment]
@kozlovic  This would probably go hand in hand with disconnecting clients and evicting the account from memory
>   We could always add an "incomplete failure count" in the account and bump this in reloadAccount() and when reaching a certain threshold stop trying (maybe disconnect clients in that case?)

File: server/accounts.go [hunk 1]
```diff
@@ -71,6 +71,7 @@ type Account struct {
 	jsLimits     *JetStreamAccountLimits
 	limits
 	expired      bool
+	incomplete   bool
```
Comment (derekcollison) [inline server/accounts.go:74]: Not for now but should turn these into flags, possibly with own mutex.

File: server/accounts.go [hunk 2]
```diff
@@ -2623,6 +2629,35 @@ func (s *Server) UpdateAccountClaims(a *Account, ac *jwt.AccountClaims) {
 			}
 		}
 	}
+
+	// if we encountered issues during this process, retry in a bit
```
Comment (derekcollison) [inline server/accounts.go:2633]: This feels like it could possibly spin no?

File: server/accounts.go [hunk 3]
```diff
@@ -2623,6 +2629,35 @@ func (s *Server) UpdateAccountClaims(a *Account, ac *jwt.AccountClaims) {
...
+		})
+	}
+}
+
+func (s *Server) reloadAccount(accName string) {
+	if !s.isRunning() {
+		return
+	}
+	if acc, ok := s.accounts.Load(accName); ok {
+		acc := acc.(*Account)
```
Comment (derekcollison) [inline server/accounts.go:2647]: Usually like to use diff variable name vs aliasing.

File: server/accounts.go (hunk 2 above)
Comment (kozlovic) [reply server/accounts.go:2633]: Like I mentioned, we could always keep track of the number of times we tried and give up after a certain amount.

File: server/accounts.go (hunk 3 above)
Comment (matthiashanel) [reply server/accounts.go]: changed first acc to a

File: server/accounts.go (hunk 2 above)
Comment (derekcollison) [reply server/accounts.go:2633]: If there is an import that fails I think that means the import is invalid but does not mean anything about the account itself. Should we be trying to reload?

Comment (matthiashanel) [reply server/accounts.go:2633]: or it means the account exporting has not been updated (to add the export) yet.

Comment (matthiashanel) [reply server/accounts.go:2633]: If this becomes an issue we can add a limit as @kozlovic said. When we do that we probably should also evict the account. (smth yet to be added)

Comment (derekcollison) [reply server/accounts.go:2633]: Can't add an import without the export. This is more the import required a token and the token expired or was revoked, or the exporting account was updated to remove or change the export.

Comment (derekcollison) [reply server/accounts.go:2633]: I don't think if an import fails the account should be evicted at all. That would mean the user can not connect at all. We would want to make it easy for the importing account to debug though.

Comment (matthiashanel) [reply server/accounts.go:2633]: If you modify two accounts (add an export and corresponding import) and push all, because the server has to process one of them first you have a high likelihood of that happening

Comment (derekcollison) [reply server/accounts.go:2633]: Hmm, good to think about the race condition for sure. My hope would be the export account update gets pushed first but I hear what you are saying.

Comment (matthiashanel) [reply server/accounts.go:2633]: > I don't think if an import fails the account should be evicted at all. That would mean the user can not connect at all. We would want to make it easy for the importing account to debug though.

The problem I see with that is noticing that smth is wrong in the first place. 
This particular error manifests itself as missing messages. 
Events associated with disconnect/reconnect are better observable than the lack of messages.

Comment (derekcollison) [reply server/accounts.go:2633]: Are we comfortable though with an account being suspended due to a bad import until said import is removed or fixed and the account re-pushed? That sounds very dangerous to me and violates path of least surprise.

Comment (matthiashanel) [reply server/accounts.go:2633]: We don't expect this condition to happen a lot, so as of now we keep trying. 
Do you think this is sufficient for now?


> Are we comfortable though with an account being suspended due to a bad import until said import is removed or  fixed and the account re-pushed? That sounds very dangerous to me and violates path of least surprise.

The account wouldn't be suspended. Clients would be able to do everything while they are connected. 
Every once in a while, they would be disconnected with a close reason account import broken.
Perhaps we could only close clients with subscriptions for the import, as well as the ones publishing to it.
However, this seems to be a discussion to be had when we get there.

Comment (derekcollison) [reply server/accounts.go:2633]: I don't think we should disconnect clients at all. We could keep trying to resolve but feels like we should be smarter. For instance we should know if an account has a broken import on it, and when we get an export update go try to reload the import account that was broke again then.

Comment (matthiashanel) [reply server/accounts.go:2633]: I take this as don't use timers then

Comment (derekcollison) [reply server/accounts.go:2633]: I would not auto try to resolve TBH, would pick concrete event that would cause us to try to resolve TBH.

Comment (matthiashanel) [reply server/accounts.go:2633]: We went down this route as we do not expect adding imports/exports happens often.
So automatically retrying once a second per initial failure appeared more straight forward than adding and maintaining data structures to track the issue.
An edge case to consider for an event driven approach is the update of the exporting account never happening. (say the import was wrong and gets removed)

Comment (derekcollison) [reply server/accounts.go:2633]: Do we have tests that show we get an import account before and export account update and produce those failures?

I don't think we should keep trying to reload or disconnect existing clients due to import failure. Just my $0.02.

Comment (kozlovic) [reply server/accounts.go:2633]: I will let Matthias comment, but he has been investigating an issue with a user that shows interest drop that has to do with import/export and accounts push. He did add tests to this PR. I do think the PR is fine as it is, the only thing we could do is stop trying after some time, but just giving up without closing clients seem wrong because then users would have the interest-loss issue which is difficult to debug as opposed to get the client disconnected.

Comment (derekcollison) [reply server/accounts.go:2633]: If I import a service and someone revokes my token, I don't think that should have all my connections cut off. I agree we need more tools for end users to be able to identify the issue/

Comment (matthiashanel) [reply server/accounts.go]: I have been contacted about the same symptom multiple times.
Because there was such a delay between dropping the subscriptions and investigating, logs turned out to be of no use.
I realized what might be going on based on narrowing down the situation to a single server + accounts with imports and then reading code.
This PR contains unit tests for fetch failing as well as unfortunate update reordering.
I'm very interested in this becoming part of the next release just so I know if I'm on the right track. 
After-all this still may be another issue altogether.

I just pushed two commits.
First one keeps track of the account names experiencing the issue.
After this was done (passed all tests) I realized that ranging over all accounts is smth that is done, and could slim down the necessary state.

So I'd say have a look at the last commit. 
Only if that does not fit the bill, have a look at the previous one.

Comment (derekcollison) [reply server/accounts.go]: ok I think we are closer. Could you squash these into 1 and ping me and will take another look all together. Thx!

File: server/accounts.go [hunk 4]
```diff
@@ -2623,6 +2632,29 @@ func (s *Server) UpdateAccountClaims(a *Account, ac *jwt.AccountClaims) {
...
+			acc := value.(*Account)
+			acc.mu.RLock()
+			incomplete := acc.incomplete
+			// Must use jwt in account or risk failing on fetch
+			// This jwt may not be the same that caused exportingAcc to be in incompleteAccExporterMap
+			claimJWT := acc.claimJWT
+			acc.mu.RUnlock()
+			if incomplete {
+				if accClaims, _, err := s.verifyAccountClaims(claimJWT); err == nil {
+					s.UpdateAccountClaims(acc, accClaims)
```
Comment (kozlovic) [inline server/accounts.go:2648]: This is the function from which we are inside. So we will possibly recurse. Do you see any problem with that?

**Review by kozlovic — CHANGES_REQUESTED** [review]
some little comment rephrasing and wondering about recursive calls.

File: server/server.go [hunk 5]
```diff
@@ -219,8 +219,8 @@ type Server struct {
 	// Websocket structure
 	websocket srvWebsocket
 
-	// exporting account name -> (incomplete) importing account
-	incompleteAccMap map[string]map[string]struct{}
+	// exporting account name the importer experienced issues
```
[the remark this answers is not in the document]
Comment (matthiashanel) [reply server/server.go:222]: fixed

File: server/accounts.go (hunk 4 above)
Comment (matthiashanel) [reply server/accounts.go:2648]: Not really.
sync.Map has no problem with it. 
We are also only recursive for cases where the account is incomplete. 
This happens at the tail end of UpdateAccountClaims, so I'm not worried about not being fully set up.
In the recursive situation, an incomplete account can turn complete and thus not be considered any longer.
At any given time I believe the number of incomplete accounts to be small.
#inclomplete accounts * # accounts.

**I guess this comes back to: is ranging over all accounts appropriate at all?**
If there is one persistently bad account every time we update the exporter we will visit all accounts.
Say the token expired, because we don't get rid of the clients or the account, we will range over all accounts.

Comment (kozlovic) [reply server/accounts.go:2648]: Valid concern and - although we do range over accounts, for instance on reload - if we plan to have many, many accounts, that could become an issue. But in order to have an impact (that is sync.map range being slow) I am guessing that it would require a LOT of accounts, no?

Comment (matthiashanel) [reply server/accounts.go:2648]: I would say we need a great many accounts. 
And then, this probably happens not that often as we remove the account from incompleteAccExporterMap.

I also want to point out that this scheme only works as long as we can continue to tolerate this situation.
update account A, fail to fetch import on account B. 
While account B has no clients, account A won't be updated.

File: server/accounts.go (hunk 2 above)
Comment (matthiashanel) [reply server/accounts.go]: Done

File: server/accounts.go [hunk 6]
```diff
@@ -2574,6 +2582,7 @@ func (s *Server) UpdateAccountClaims(a *Account, ac *jwt.AccountClaims) {
 		}
 	}
 	a.defaultPerms = buildPermissionsFromJwt(&ac.DefaultPermissions)
+	a.incomplete = incomplete
 	a.mu.Unlock()
```
Comment (derekcollison) [inline server/accounts.go:2586]: nit but maybe here check incomplete and do the ExporterMap assignment once here.

File: server/accounts.go [hunk 7]
```diff
@@ -2623,6 +2632,30 @@ func (s *Server) UpdateAccountClaims(a *Account, ac *jwt.AccountClaims) {
 			}
 		}
 	}
+
+	if _, ok := s.incompleteAccExporterMap.Load(old.Name); ok {
+		s.incompleteAccExporterMap.Delete(old.Name)
+		s.accounts.Range(func(key, value interface{}) bool {
```
Comment (derekcollison) [inline server/accounts.go:2638]: Hard to follow really what is going on here, comment might be good.

**Review by derekcollison — APPROVED** [review]
Some minor comments but LGTM.

File: server/accounts.go (hunk 6 above)
Comment (matthiashanel) [reply server/accounts.go:2586]: Not sure I follow. When we iterated over all imports and did NOT run into an error, it does not matter what value a.incomplete had. At this point it is complete.

Since each import can refer to a different export there is not really such a thing as once.
I assume assigning exportermap once is about setting it together with a.incomplete. (will change that)

File: server/accounts.go (hunk 7 above)
Comment (matthiashanel) [reply server/accounts.go:2638]: added

File: server/accounts.go [hunk 8]
```diff
@@ -2647,6 +2647,10 @@ func (s *Server) UpdateAccountClaims(a *Account, ac *jwt.AccountClaims) {
 			if incomplete && name != old.Name {
 				if accClaims, _, err := s.verifyAccountClaims(claimJWT); err == nil {
 					s.UpdateAccountClaims(acc, accClaims)
+					// old.Name was deleted before ranging over accounts
```
Comment (kozlovic) [inline server/accounts.go:2650]: I am uneasy with this approach of doing it in place (as opposed to timer based). Say that we had A import from B that import from C, and all this chain is broken (meaning processing of A import failed, and B import failed).
Say that we are processing A at the moment (so old.Name == A). There is no guarantee of ordering doing the accounts.Range(). So say that the first we get is "B" in the Range() and it is incomplete (and name != A) then we recursively call this function. Now we are back at line 2636 and we have "B" marked as incomplete. We range and now say that the first in the Range() is "A". It is still incomplete and we now recursively call this function again, and we happen to going from "A" being given first in Range(), to "B", would that not end-up panic'ing with a stack overflow?

Comment (matthiashanel) [reply server/accounts.go:2660]: permanently broken circular imports did the trick. fixed

File: server/jwt_test.go [hunk 9]
```diff
@@ -1911,6 +1911,110 @@ func TestAccountURLResolverFetchFailurePushReorder(t *testing.T) {
...
+	ncB := natsConnect(t, sA.ClientURL(), createUserCreds(t, nil, expkp))
+	defer ncB.Close()
+	sysc := natsConnect(t, sA.ClientURL(), createUserCreds(t, nil, syskp))
+	defer sysc.Close()
+	// push accounts
+	natsPub(t, sysc, fmt.Sprintf(accUpdateEventSubj, imppub), []byte(impjwt))
+	natsPub(t, sysc, fmt.Sprintf(accUpdateEventSubj, exppub), []byte(expjwt))
+	sysc.Flush()
+	importErrCnt := 0
+	for {
```
Comment (kozlovic) [inline server/jwt_test.go:2002]: This is going to be long test (2 or 4 seconds, etc..). Replace with this:
```
get_errors_loop:
	for {
		select {
		case line := <-l.dbgCh:
			if strings.HasPrefix(line, "Error adding stream import to account") {
				fmt.Printf("@@IK: here: %q\n", line)
				importErrCnt++
				if importErrCnt == 6 {
					break get_errors_loop
				}
			}
		case <-time.After(2 * time.Second):
			t.Fatal("Did not get all 6 debug traces")
		}
	}
	// Make sure we don't get more..
	for {
		select {
		case line := <-l.dbgCh:
			if strings.HasPrefix(line, "Error adding stream import to account") {
				t.Fatalf("Got more error lines than expected: %q", line)
			}
		case <-time.After(250 * time.Millisecond):
			// OK
			return
		}
	}
```

Comment (kozlovic) [reply server/jwt_test.go:2002]: or, as discussed, make sure you create the timer outside of for loop and reduce to 500ms.

Comment (matthiashanel) [reply server/jwt_test.go:2004]: done
